packages/opentelemetry-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,148 @@
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  get state(): HookState {
+    return this._hookState;
+  }
+
+  get plugins(): PluginType {
+    return this._plugins;
+  }
+
+  /**
+   * Loads a list of plugins. Each plugin module should implement the core
+   * {@link Plugin} interface and export an instance named as 'plugin'. This
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param modulesToPatch A list of plugins.
+   */
+  loadPlugins(modulesToPatch: string[]): PluginLoader {
+    if (this._hookState === HookState.UNINITIALIZED) {
+      const plugins = modulesToPatch.reduce(
+        (plugins: PluginNames, moduleName: string) => {
+          plugins[moduleName] = utils.defaultPackageName(moduleName);
+          return plugins;
+        },
+        {} as PluginNames
+      );
+      const modulesToHook = Object.keys(plugins);
+      // Do not hook require when no module is provided. In this case it is
+      // not necessary. With skipping this step we lower our footprint in
+      // customer applications and require-in-the-middle won't show up in CPU
+      // frames.
+      if (modulesToHook.length === 0) {
+        this._hookState = HookState.DISABLED;
+        return this;
+      }
+
+      // Enable the require hook.
+      hook(modulesToHook, (exports, name, baseDir) => {
+        if (this._hookState !== HookState.ENABLED) return exports;
+
+        const moduleName = plugins[name];
+        // Get the module version.
+        const version = utils.getPackageVersion(this.logger, baseDir as string);
+        this.logger.info(
+          `PluginLoader#loadPlugins: trying loading ${name}.${version}`
+        );
+        if (!version) return exports;
+
+        this.logger.debug(
+          `PluginLoader#loadPlugins: applying patch to ${name}@${version} using ${moduleName} module`
+        );
+
+        // Expecting a plugin from module;
+        try {
+          // tslint:disable-next-line:no-any
+          const plugin: Plugin<any> = require(moduleName).plugin;
+          this._plugins.push(plugin);
+          // Enable each supported plugin.
+          return plugin.enable(exports, this.tracer);
+        } catch (e) {
+          this.logger.error(
+            `PluginLoader#loadPlugins: could not load plugin ${moduleName} of module ${name}. Error: ${e.message}`
+          );
+          return exports;
+        }
+      });
+      this._hookState = HookState.ENABLED;
+    } else if (this._hookState === HookState.DISABLED) {
+      throw new Error(
+        'PluginLoader#loadPlugins: Currently cannot re-enable plugin loader.'
+      );
+    } else {
+      throw new Error(
+        'PluginLoader#loadPlugins: Plugin loader already enabled.'
+      );
+    }
+    return this;
+  }
+
+  /** Unloads plugins. */
+  unloadPlugins(): PluginLoader {
```

Comment:
`unloadPlugins` is redundant since it's a plugin loader. Just `unload` would be enough.

Reply:
sure, done